src/Npgsql/TypeHandlers/CompositeHandlers/CompositeHandler.cs

```diff
@@ -97,31 +101,133 @@ public override int ValidateAndGetLength(T value, ref NpgsqlLengthCache? lengthC
+
+                _memberHandlers = CreateMemberHandlers(pgType, _typeMapper, _nameTranslator);
+                _constructorHandler = CreateConstructorHandler(pgType, _typeMapper, _nameTranslator);
+                _constructor = _constructorHandler is null
+                    ? Expression
+                        .Lambda<Func<T>>(Expression.New(typeof(T)))
+                        .Compile()
+                    : null;
+            }
+        }
+
+        static CompositeConstructorHandler<T>? CreateConstructorHandler(PostgresCompositeType pgType, ConnectorTypeMapper typeMapper, INpgsqlNameTranslator nameTranslator)
         {
+            var pgFields = pgType.Fields;
+
             var clrType = typeof(T);
+            var clrParameterHandlerType = typeof(CompositeParameterHandler<>);
+
+            ConstructorInfo? clrDefaultConstructor = null;
+
+            foreach (var clrConstructor in clrType.GetConstructors())
+            {
+                var clrParameters = clrConstructor.GetParameters();
+                if (clrParameters.Length != pgFields.Count)
+                {
+                    if (clrParameters.Length == 0)
+                        clrDefaultConstructor = clrConstructor;
+
+                    continue;
+                }
+
+                var clrParameterHandlerCount = 0;
+                var clrParameterIndexes = new int[pgFields.Count];
+
+                for (var pgFieldIndex = 0; pgFieldIndex < pgFields.Count; ++pgFieldIndex)
+                    clrParameterIndexes[pgFieldIndex] = -1;
+
+                foreach (var clrParameter in clrParameters)
+                {
+                    var attr = clrParameter.GetCustomAttribute<PgNameAttribute>();
+                    var name = attr?.PgName ?? (clrParameter.Name is string clrName ? nameTranslator.TranslateMemberName(clrName) : null);
+                    if (name is null)
+                        break;
+
+                    for (var pgFieldIndex = pgFields.Count - 1; pgFieldIndex >= 0; --pgFieldIndex)
+                    {
+                        var pgField = pgFields[pgFieldIndex];
+                        if (pgField.Name != name)
+                            continue;
+
+                        if (clrParameterIndexes[pgFieldIndex] != -1)
+                            throw new AmbiguousMatchException($"Multiple constructor parameters are mapped to the '{pgField.Name}' field.");
+
+                        if (!typeMapper.TryGetByOID(pgField.Type.OID, out var handler))
+                            throw new NpgsqlException($"PostgreSQL composite type {pgType.DisplayName} has field {pgField.Type.DisplayName} with an unknown type (OID = {pgField.Type.OID}).");
+
+                        clrParameterHandlerCount++;
+                        clrParameterIndexes[pgFieldIndex] = clrParameter.Position;
+
+                        break;
+                    }
+                }
+
+                if (clrParameterHandlerCount < pgFields.Count)
```

Comment:
`clrParameterIndexes` is an `int` array, so there are no nulls.

Reply:
OK, we can check for -1 instead. Or if you accept [my suggestion](https://github.com/npgsql/npgsql/pull/2523#discussion_r327612586) to have an array of clrParameters instead of positions then it's nulls.

Reply:
Changed to be an array of `ParameterInfo`, thanks, it looks better now. Consider using LINQ I'm strongly disagree because there is no need to loop once more through array when it's possible to count handlers in the existing loop.

Reply:
We should really distinguish between perf-sensitive code and other code. It's totally fine, in a function that runs once, to iterate twice over a small list instead of keeping another state variable - the perf impact is zero and the readability/logic is clearer. But it's not critical.